src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -105,18 +105,19 @@ private void ParseRequestLine(TRequestHandler handler, ReadOnlySpan<byte> reques
         offset++;
 
         // Find end of path and if path is encoded
-        for (; (uint)offset < (uint)requestLine.Length; offset++)
+        var index = requestLine.Slice(offset).IndexOfAny(ByteSpace, ByteQuestionMark, BytePercentage);
+        if (index >= 0)
         {
-            ch = requestLine[offset];
-            if (ch == ByteSpace || ch == ByteQuestionMark)
-            {
-                // End of path
-                break;
-            }
-            else if (ch == BytePercentage)
+            if (requestLine[offset + index] == BytePercentage)
             {
                 pathEncoded = true;
+                offset += index;
+                // Found an encoded character, now just search for end of path
```

Comment:
Does this mean there can be at most one encoded character?

Reply:
No, it sets a bool so later code can process it differently. We're just searching for the segment that contains the path.